pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,383 @@
+			ac.SetHeartBeater(ac.pool.heartBeatCreator())
+
+			break
+		}
+
+	}, func(r interface{}) {
+		log.DefaultLogger.Warnf("[connpool] reconnect failed, %v, host: %v", r, ac.host.Host.AddressString())
+	})
+
+}
+
+// removeFromPool removes this client from connection pool
+func (ac *activeClient) removeFromPool() {
+	p := ac.pool
+	p.clientMux.Lock()
+	defer p.clientMux.Unlock()
+
+	p.client = nil
+}
+
+// SetHeartBeater set the heart beat for an active client
+func (ac *activeClient) SetHeartBeater(hb KeepAlive) {
+	// clear the previous keepAlive
+	if ac.keepAlive != nil && ac.keepAlive.keepAlive != nil {
+		ac.keepAlive.keepAlive.Stop()
+		ac.keepAlive = nil
+	}
+
+	ac.keepAlive = &keepAliveListener{
+		keepAlive: hb,
+		conn:      ac.host.Connection,
+	}
+
+	// this should be equal to
+	// ac.codecClient.AddConnectionEventListener(ac.keepAlive)
+	ac.host.Connection.AddConnectionEventListener(ac.keepAlive)
+}
+
+// keepAliveListener is a types.ConnectionEventListener
+type keepAliveListener struct {
+	keepAlive KeepAlive
+	conn      api.Connection
+}
+
+// OnEvent impl types.ConnectionEventListener
+func (l *keepAliveListener) OnEvent(event api.ConnectionEvent) {
+	if event == api.OnReadTimeout && l.keepAlive != nil {
+		heartbeatFailCreator := func() {
+			l.conn.Close(api.NoFlush, api.OnReadErrClose)
+		}
+
+		// TODO, whether error should be handled
+		l.conn.Write(buffer.NewIoBufferBytes(l.keepAlive.GetKeepAliveData(heartbeatFailCreator)))
+	}
+}
+
+type KeepAlive interface {
+	Stop()
+
+	GetKeepAliveData(failCallback func()) []byte
+}
+
+type Connection interface {
+	Write(buf ...buffer.IoBuffer) error
+	Destroy()
+	Available() bool
+}
+
+// NewConn returns a simplified connpool
+func NewConn(hostAddr string, reconnectTryTimes int, heartBeatCreator func() KeepAlive, readFilters []api.ReadFilter, autoReconnectWhenClose bool) Connection {
+	// use host addr as cluster name, for the count of metrics
+	cl := basicCluster(hostAddr, []string{hostAddr})
+	host := cluster.NewSimpleHost(cl.Hosts[0], cluster.NewCluster(cl).Snapshot().ClusterInfo())
+
+	// if user configure this to -1, then retry is unlimited
+	if reconnectTryTimes == -1 {
+		reconnectTryTimes = math.MaxInt32
+	}
+
+	p := &connpool{
+		host:                   host,
+		heartBeatCreator:       heartBeatCreator,
+		autoReconnectWhenClose: autoReconnectWhenClose,
+		reconnTryTimes:         reconnectTryTimes,
+		readFilters:            readFilters,
+	}
+
+	// trigger the client generation
+	p.GetActiveClient(context.Background())
+
+	return p
+}
+
+// write to client
+func (p *connpool) Write(buf ...buffer.IoBuffer) error {
+	cli, reason := p.GetActiveClient(context.Background())
+	if reason != "" {
+		return errors.New("get client failed" + string(reason))
+	}
+
+	// use atomic to load pointer, avoid partial pointer read
+	h := (*types.CreateConnectionData)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&cli.host))))
+
+	return h.Connection.Write(buf...)
```

Comment:
maybe panic if connection is nil？

Reply:
CreateConnection ensures the Connection field is never nil
if ac.host is not nil, then the Connection is not nil